Add quadkey conversion to xy coordinates in the Bingtiles grid generator

tm-abby-moreno opened the pull request:
This PR is based on the replies in discussion #90 

Added a boolean parameter `add_xy_cols` in the `class BingTileGridGenerator` whether user would want to convert quadkey to xy coordinates. The newly created function `quad_to_xy` is based on this [stackexchange](https://gis.stackexchange.com/questions/206202/how-can-i-convert-bings-quadtree-tile-addresses-to-zxy-tile-addresses-in-pyth). 

I tested this on `zoom_level = 12` and speed is still the same! I then tested it on `zoom_level = 16` because this was the use case for CI and the baseline (without quadkey conversion) is ~32 mins. and when there is quadkey conversion to xy it still runs for 30 mins. TLDR: the things that I added did not impact the runtime 😄 I also added to the tutorial and fixed some typos in other comments.

jtmiclat commented on geowrangler/grids.py:
```diff
@@ -200,21 +202,23 @@ def __init__(
+        """Get the intersecting tiles with polygon for a zoom level. Polygon should be in EPSG:4326"""
         x_min, y_min, x_max, y_max = polygon.bounds
         tiles = (
             (self.tms.quadkey(tile), self.tile_to_polygon(tile))
```
I would prefer you extract xyz from tile object here rather than recomputing it. 

You can access the x,y,z from a tile via `tile.x`, `tile.y`, `tile.z`

You can update the following line to  

```python
   tiles = (
            (self.tms.quadkey(tile), self.tile_to_polygon(tile), tile.x, tile,y, tile.z) # add x, y, z
            for tile in self.tms.tiles(x_min, y_min, x_max, y_max, self.zoom_level)
        )
        # Return dict to make it easier to deduplicate
        if filter:
            tiles = {qk: (geom,x,y,z) for qk, geom, x, y, z,  in tiles if polygon.intersects(geom)} # upadte returned type to a tuple
        else:
            tiles = {qk:  (geom,x,y,z) for qk, geom, x, y, z in tiles} # update returned type to a tuple
        return tiles
 ```

later down you would replace
```python
quadkey, geom = zip(*((k, v) for k, v in tiles.items()))
```
with 
```python
quadkey, geom_xyz, = zip(*((k, v) for k, v in tiles.items()))
geom, x, y, z = zip(*geom_xyz)
```

then you can do
```python
df['x'] = list(x)
```

butchtm replied:
Somewhat related but I'm also writing a [parallelized version of the filtering part ](https://github.com/thinkingmachines/geowrangler/blob/4b6292354c16b8676f868b7d69e505930730687b/geowrangler/grids.py#L342) because based on my profiling on bing tile gridding for large areas (e.g. Indonesia), this is where the process bottlenecked.

jtmiclat reviewed (changes requested):
Pls add tests as well

tm-abby-moreno commented:
Updated the code to call xyz directly and added a test @jtmiclat @butchtm :) 

I took some time as I was testing it out. For `zoom_level = 15` it runs for only ~13 mins. but when I tried it the first time for `zoom_level = 16` it ran for ~50 mins. But I tried running again just this morning and it ran for ~32 mins. Just wanted to flag this! Not exactly sure what the problem is. Note: I always test it on 16 because this is the use-case for Josh in CI. 

**re: failing pre-commit test** how do I fix this?

jtmiclat commented:
@tm-abby-moreno. You can ignore the pre-commit test failing. I took a look and the issue isnt with the code you changed but a dependency issue in the tests. 

Addressed that issue in a separate PR: 
https://github.com/thinkingmachines/geowrangler/pull/208

jtmiclat commented on geowrangler/grids.py:
```diff
@@ -240,14 +246,34 @@ def generate_grid(self: BingTileGridGenerator, gdf: GeoDataFrame) -> DataFrame:
+
+    result = {"quadkey": list(quadkey)}
+
+    if self.add_xyz_cols:
```
small nitpick but this can simplified to
```python
    if self.add_xyz_cols:
        result["x"] = [t.x for t in tile]
        result["y"] = [t.y for t in tile]
        result["z"] = [t.z for t in tile]

    if self.return_geometry:
        tiles_df = GeoDataFrame(
            result,
            geometry=list(geom),
            crs="epsg:4326",
        )
        tiles_df = tiles_df.to_crs(gdf.crs)
    else:
        tiles_df = DataFrame(result)

    return tiles_df
```

jtmiclat reviewed (approved):
Approving with a small nitpick! Feel free to address the nitpick or not haha
